### PR TITLE
#8711: Handle disabled logging in 'caplog.set_level' and 'caplog.at_level' 

### LINKED ISSUE #8711: Handle logging.disable() in caplog.set_level and caplog.at_level
### What's the problem this feature will solve?

Currently, caplog will not catch logs if `logging.disable(LEVEL)` was called somewhere, in my case it's called in my django
settings for test runs.

### Describe the solution you'd like

`caplog.at_level` and `caplog.set_level` should both take a new arg called `force_enable` that tells `caplog` to set the disable setting to one level below whatever level was requested to caplog. 
Or it should force enable the logging level if `rootLogger.manager.disable >= level` so that logging will always
be enabled for the level a test is trying to capture.

This will allow caplog to capture logs at the desired level even if they were disabled. 

e.g.

```python
# settings/testing.py
# disables all logs
import logging
logging.disable(logging.CRITICAL)


# test_file.py
import logging


def test_stuff(caplog):
    # now everything *below* WARNING will be disabled instead of everything
    caplog.set_level(logging.WARNING)
    logging.error("Sierra 117")
    assert len(caplog.records) == 1
    assert "117" in caplog.text
```

The test will currently fail because caplog doesn't override `logging.disable` which is a separate setting from log level.

My proposal is to add a call to `logging.disable(max(level-10, 0))` inside both `set_level` and `at_level`. The `-10` means disable everything *below* the requested level for `caplog`. We set a minimum of `0` because that's the lowest log level.

This can be reset by adding an attribute to keep track of the initial disabled level, similar to the one added to fix https://github.com/pytest-dev/pytest/issues/7133.

### Alternative Solutions

This is my current workaround in one of my tests

```
with caplog.at_level(logging.WARNING):
    # logging.disable needs to be called because caplog doesn't handle disabled logging.
    logging.disable(logging.WARNING)
    response = ah_client.get("/api/redacted/")
    # re-disable all logs.
    logging.disable(logging.CRITICAL)
```

### Additional context

I have already written the code and tests and will submit a PR for review or link to the branch
of my `pytest` fork in the comments of this issue.

### PR DESCRIPTION
**Pull request by alexlambson**
Forces requested `caplog` logging levels to be enabled if they were disabled via `logging.disable()`

`[attr-defined]` mypy error ignored in `logging.py` because there were existing errors with the imports
and `loggin.Logger.manager` is an attr set at runtime. Since it's in the standard lib I can't really fix that.

Ignored an attr-defined error in `src/_pytest/config/__init__.py` because the re-export is necessary.

Closes #8711

### CODE REVIEW

File: src/_pytest/logging.py
Comment (alexlambson) [inline src/_pytest/logging.py:456]: Linter doesn't like runtime attributes, so the ignore is needed to the best of my knowledge.

File: src/_pytest/logging.py
Comment (alexlambson) [inline src/_pytest/logging.py:467]: Each level is `10` away from other levels.

File: src/_pytest/logging.py
Comment (alexlambson) [inline src/_pytest/logging.py:493]: We could remove this magic for a parameter if magic is frowned upon.

**Comment by alexlambson** [comment]
Just bumping / following up on this because it's popped up for me again in a project.

File: src/_pytest/logging.py
Comment (andrewdotn) [inline src/_pytest/logging.py:444]:
```suggestion
        Only enables logging levels equal to and higher than the requested ``level``.
```

File: testing/logging/test_fixture.py
Comment (andrewdotn) [inline testing/logging/test_fixture.py:107]: Since this needs to run, I’d make it a little fixture, so that logging gets restored even if the test exits early due to a failed assert.

File: src/_pytest/logging.py
Comment (andrewdotn) [inline src/_pytest/logging.py:503]:
```suggestion
        Will enable the requested logging level if it was disabled via :meth:`logging.disable`.
```

File: src/_pytest/logging.py
Comment (andrewdotn) [inline src/_pytest/logging.py:470]: I’d recommend using `elif` instead of `return` and consolidating the three `return` statements to one.

```suggestion
        if not isinstance(level, int):
            logging.disable(logging.NOTSET)
        elif not logger_obj.isEnabledFor(level):
            # Each level is `10` away from other levels.
            # https://docs.python.org/3/library/logging.html#logging-levels
            disable_level = max(level - 10, logging.NOTSET)
            logging.disable(disable_level)        
        return original_disable_level
```

Is the logic still correct there?

File: src/_pytest/logging.py
Comment (andrewdotn) [inline src/_pytest/logging.py:479]:
```suggestion
        Will enable the requested logging level if it was disabled via :meth:`logging.disable`.
```

**Review by andrewdotn — COMMENTED** [review]
Thanks for sending this in! I’m new here, so I hope my comments are helpful.

File: src/_pytest/logging.py
Comment (alexlambson) [reply src/_pytest/logging.py:444]: Well that's embarrassing. I'll fix it along with all other typos.

File: testing/logging/test_fixture.py
Comment (alexlambson) [reply testing/logging/test_fixture.py:107]: Oh that's a good idea. Good catch. I'll try to figure that out

File: src/_pytest/logging.py
Comment (alexlambson) [reply src/_pytest/logging.py:503]: How did I make so many typos :'(

File: src/_pytest/logging.py
Comment (alexlambson) [reply src/_pytest/logging.py:470]: I am partial to return early, return often, but if this is the preferred style then I'll update my code to match. I'll make the change and make sure it still works.

File: src/_pytest/logging.py
Comment (alexlambson) [reply src/_pytest/logging.py:479]: I guess my "t" finger was tired that day.

**Comment by alexlambson** [comment]
> 
> 
> Thanks for sending this in! I’m new here, so I hope my comments are helpful.

Thanks for the review. I'll fix these in the morning :)

File: src/_pytest/logging.py
Comment (alexlambson) [reply src/_pytest/logging.py]: fixed

File: testing/logging/test_fixture.py
Comment (alexlambson) [reply testing/logging/test_fixture.py]: fixed

File: src/_pytest/logging.py
Comment (alexlambson) [reply src/_pytest/logging.py]: fixed

File: src/_pytest/logging.py
Comment (alexlambson) [reply src/_pytest/logging.py]: fixed

File: src/_pytest/logging.py
Comment (alexlambson) [reply src/_pytest/logging.py]: fixed

**Comment by alexlambson** [comment]
The code coverage seems to not look into those string-based tests. Most of the untested lines are actually tested.

**Comment by alexlambson** [comment]
Fixed test coverage

**Comment by blerrgh** [comment]
What happened to this? Would be a super handy feature to have.

**Comment by alexlambson** [comment]
Mainly just waiting for approval

File: src/_pytest/logging.py
Comment (nicoddemus) [inline src/_pytest/logging.py:457]:
```suggestion
        """Enable the desired logging level if the global level was disabled via ``logging.disabled``.
```

**Review by nicoddemus — CHANGES_REQUESTED** [review]
Hi @alexlambson,

Really sorry this fell through the tracks and got forgotten for so long, I appreciate the recent pings.

I left a few minor comments, the PR is really well written, including the tests.

If you could just rebase it and fix the comments I left, I will be glad to get this merged. 

:+1:

**Review by nicoddemus — CHANGES_REQUESTED** [review]
Hi @alexlambson,

Really sorry this fell through the tracks and got forgotten for so long, I appreciate the recent pings.

I left a few minor comments, the PR is really well written, including the tests.

If you could just rebase it and fix the comments I left, I will be glad to get this merged. 

:+1:

File: src/_pytest/logging.py
Comment (alexlambson) [reply src/_pytest/logging.py]: fixed

File: src/_pytest/logging.py
Comment (alexlambson) [reply src/_pytest/logging.py]: switched

File: src/_pytest/logging.py
Comment (alexlambson) [reply src/_pytest/logging.py]: Thank you, this knowledge will save me some headache at work.

**Comment by alexlambson** [comment]
No worries @nicoddemus, pytest is a busy project.

**Review by nicoddemus — APPROVED** [review]
Thanks a lot @alexlambson!

**Pull request merged**
